include/pipert/Protocol.h

```diff
@@ -0,0 +1,37 @@
+#ifndef _PROTOCOLHANDSHAKE_H_
+#define _PROTOCOLHANDSHAKE_H_
+
+#include "pipert/UDPConnection.h"
+
+#include <cstddef>
+
+namespace pipert {
+
+template <class T>
+class Protocol {
+ public:
+  static char HandshakeSuccessMsg[];
+  static char HandshakeErrorMsg[];
+
+  Protocol(const UDPConnection& connection);
+  ~Protocol() = default;
+
+  void SenderSideHandshake();
+  void ReceiverSideHandshake();
+
+ private:
+  UDPConnection connection_;
```

Comment:
If you replace this to a pointer (and you make sure, that you allocate and release memory correctly, you can get rid of the UDPConnection being public API.

Reply:
...as in the header, you are not using this to anything else.

Reply:
You could also have a ProtocolImpl, but I think it is overkill and this trick will do.